mono/profiler/proflog.c

```diff
@@ -487,8 +488,102 @@ ign_res (int G_GNUC_UNUSED unused, ...)
 {
 }
 
-#define ENTER_LOG(LB, STR)
-#define EXIT_LOG(LB)
+/*
+ * These macros create a scope to avoid leaking the buffer returned
+ * from ensure_logbuf () as it may have been invalidated by a GC
+ * thread during STW.
+ */
+
+#define ENTER_LOG \
+	do { \
+		buffer_lock ()
+
+#define EXIT_LOG \
+		buffer_unlock (); \
+	} while (0)
+
+static volatile gint32 buffer_rwlock_count;
+static volatile gpointer buffer_rwlock_exclusive;
+
+// Can be used recursively.
+static void
+buffer_lock (void)
+{
+	gpointer tid = (gpointer) thread_id ();
+	gpointer excl = InterlockedReadPointer (&buffer_rwlock_exclusive);
+
+	/*
+	 * If the thread holding the exclusive lock tries to modify the
+	 * reader count, just make it a no-op. This way, we also avoid
+	 * invoking the GC safe point macros below, which could break if
+	 * done from a thread that is currently the initiator of STW.
+	 *
+	 * In other words, we rely on the fact that the GC thread takes
+	 * the exclusive lock in the gc_event () callback when the world
+	 * is about to stop.
+	 */
+	if (excl == tid)
+		goto done;
+
+	MONO_ENTER_GC_SAFE;
+
+	while ((excl = InterlockedReadPointer (&buffer_rwlock_exclusive)) && excl != tid)
+		mono_thread_info_yield ();
+
+	InterlockedIncrement (&buffer_rwlock_count);
+
+	MONO_EXIT_GC_SAFE;
+
+done:
+	mono_memory_barrier ();
+}
+
+static void
+buffer_unlock (void)
+{
+	mono_memory_barrier ();
+
+	// See the comment in buffer_lock ().
+	if (InterlockedReadPointer (&buffer_rwlock_exclusive) == (gpointer) thread_id ())
+		return;
+
+	g_assert (InterlockedRead (&buffer_rwlock_count) && "Why are we trying to decrement a zero reader count?");
+
+	InterlockedDecrement (&buffer_rwlock_count);
+}
+
+// Cannot be used recursively.
+static void
+buffer_lock_excl (void)
+{
+	gpointer tid = (gpointer) thread_id ();
+
+	g_assert (InterlockedReadPointer (&buffer_rwlock_exclusive) != tid && "Why are we taking the exclusive lock twice?");
+
+	MONO_ENTER_GC_SAFE;
+
+	while (InterlockedCompareExchangePointer (&buffer_rwlock_exclusive, tid, 0))
```

Comment:
so you went with a dekker lock. definitely better. it bias exclusive over shared, which is what we want.

I think it's correct.